### PR TITLE
Add the option of monitoring the residual history from linear solvers

### PR DESCRIPTION
**Pull request by njansson**
This adds the option of enabling logging (real time) of residual histories from each linear solver.

The feature is enabled per solver block in the case file, by setting `monitor` to `true`, e.g:

```json
        "pressure_solver": {
            "type": "gmres",
            "preconditioner": "hsmg",
            "projection_space_size": 4,
            "absolute_tolerance": 1e-7,
            "max_iterations": 800,
            "monitor": true     <--- This enables the feature
        },
```

### CODE REVIEW

File: src/scalar/scalar_scheme.f90 [hunk 1]
```diff
@@ -554,14 +558,17 @@ end subroutine scalar_scheme_validate
...
   !! @note Currently only supporting Krylov solvers
-  subroutine scalar_scheme_solver_factory(ksp, n, solver, max_iter, abstol)
+  subroutine scalar_scheme_solver_factory(ksp, n, solver, max_iter, &
+                                          abstol, monitor)
     class(ksp_t), allocatable, target, intent(inout) :: ksp
     integer, intent(in), value :: n
     integer, intent(in) :: max_iter
     character(len=*), intent(in) :: solver
     real(kind=rp) :: abstol
+    logical monitor
```
Comment (timfelle) [inline src/scalar/scalar_scheme.f90:568]: Intentions are not clear for these variables.

File: src/krylov/krylov_fctry.f90 [hunk 2]
```diff
@@ -73,14 +73,16 @@
...
   module subroutine krylov_solver_factory(object, n, type_name, &
-       max_iter, abstol, M)
+       max_iter, abstol, M, monitor)
     class(ksp_t), allocatable, target, intent(inout) :: object
     integer, intent(in), value :: n
     character(len=*), intent(in) :: type_name
     integer, intent(in) :: max_iter
     real(kind=rp), optional :: abstol
     class(pc_t), optional, intent(inout), target :: M
+    logical, optional :: monitor
```
Comment (timfelle) [inline src/krylov/krylov_fctry.f90:85]: What is the intent of this variable

File: src/krylov/krylov.f90 [hunk 3]
```diff
@@ -178,14 +186,16 @@ end subroutine ksp_t_free
...
      module subroutine krylov_solver_factory(object, n, type_name, &
-          max_iter, abstol, M)
+          max_iter, abstol, M, monitor)
        class(ksp_t), allocatable, target, intent(inout) :: object
        integer, intent(in), value :: n
        character(len=*), intent(in) :: type_name
        integer, intent(in) :: max_iter
        real(kind=rp), optional :: abstol
        class(pc_t), optional, intent(inout), target :: M
+       logical, optional :: monitor
```
Comment (timfelle) [inline src/krylov/krylov.f90:198]: intent

File: src/krylov/krylov.f90 [hunk 4]
```diff
@@ -202,12 +212,13 @@ end subroutine krylov_solver_destroy
   !! @param rel_tol Relative tolarance for converence.
   !! @param rel_tol Absolute tolarance for converence.
   !! @param M The preconditioner.
-  subroutine krylov_init(this, max_iter, rel_tol, abs_tol, M)
+  subroutine krylov_init(this, max_iter, rel_tol, abs_tol, M, monitor)
     class(ksp_t), target, intent(inout) :: this
     integer, intent(in) :: max_iter
```
Comment (timfelle) [inline src/krylov/krylov.f90:217]: intent

File: src/krylov/krylov.f90 [hunk 5]
```diff
@@ -202,12 +212,13 @@ end subroutine krylov_solver_destroy
...
     real(kind=rp), optional, intent(in) :: rel_tol
     real(kind=rp), optional, intent(in) :: abs_tol
     class(pc_t), optional, target, intent(in) :: M
+    logical, optional :: monitor
```
Comment (njansson) [inline src/krylov/krylov.f90:221]:
```suggestion
    logical, optional, intent(in) :: monitor
```

File: src/krylov/krylov_fctry.f90 (hunk 2 above)
Comment (njansson) [inline src/krylov/krylov_fctry.f90:85]:
```suggestion
    logical, optional, intent(in) :: monitor
```

File: src/scalar/scalar_scheme.f90 (hunk 1 above)
Comment (njansson) [inline src/scalar/scalar_scheme.f90:568]:
```suggestion
    logical, intent(in) :: monitor
```

File: src/krylov/krylov.f90 (hunk 3 above)
Comment (njansson) [inline src/krylov/krylov.f90:198]:
```suggestion
       logical, optional, intent(in) :: monitor
```

File: src/krylov/krylov.f90 [hunk 6]
```diff
@@ -267,4 +284,48 @@ subroutine krylov_set_pc(this, M)
...
+    if (this%monitor) then
+       call neko_log%end()
+       call neko_log%end_section()
+       call neko_log%newline()
+    end if
+  end subroutine krylov_monitor_stop
+
+  
+  !> Monitor iteration
+  subroutine krylov_monitor_iter(this, iter, rnorm)
```
Comment (MartinKarp) [inline src/krylov/krylov.f90:318]: So we go with this in the log always, are you planning to add the option to output to csv or similar? I think this is really nice.

File: src/krylov/bcknd/device/gmres_device.F90 [hunk 7]
```diff
@@ -131,14 +131,15 @@ end function device_gmres_part2
...
-       rel_tol, abs_tol)
+       rel_tol, abs_tol, monitor)
     class(gmres_device_t), target, intent(inout) :: this
     integer, intent(in) :: n
     integer, intent(in) :: max_iter
     class(pc_t), optional, intent(inout), target :: M
     integer, optional, intent(inout) :: m_restart
     real(kind=rp), optional, intent(inout) :: rel_tol
     real(kind=rp), optional, intent(inout) :: abs_tol
+    logical, optional :: monitor
```
Comment (timfelle) [inline src/krylov/bcknd/device/gmres_device.F90:142]:
```suggestion
    logical, optional, intent(in) :: monitor
```

File: src/krylov/bcknd/device/pipecg_device.F90 [hunk 8]
```diff
@@ -168,13 +168,14 @@ subroutine device_cg_update_xp(x_d, p_d, u_d_d, alpha, beta, p_cur, p_space, n)
...
   !> Initialise a pipelined PCG solver
-  subroutine pipecg_device_init(this, n, max_iter, M, rel_tol, abs_tol)
+  subroutine pipecg_device_init(this, n, max_iter, M, rel_tol, abs_tol, monitor)
     class(pipecg_device_t), target, intent(inout) :: this
     class(pc_t), optional, intent(inout), target :: M
     integer, intent(in) :: n
     integer, intent(in) :: max_iter
     real(kind=rp), optional, intent(inout) :: rel_tol
     real(kind=rp), optional, intent(inout) :: abs_tol
+    logical, optional :: monitor
```
Comment (timfelle) [inline src/krylov/bcknd/device/pipecg_device.F90:178]:
```suggestion
    logical, optional, intent(in) :: monitor
```

File: src/krylov/bcknd/sx/cg_sx.f90 [hunk 9]
```diff
@@ -60,13 +60,14 @@ module cg_sx
...
   !> Initialise a standard PCG solver
-  subroutine sx_cg_init(this, n, max_iter, M, rel_tol, abs_tol)
+  subroutine sx_cg_init(this, n, max_iter, M, rel_tol, abs_tol, monitor)
     class(sx_cg_t), intent(inout) :: this
     class(pc_t), optional, intent(inout), target :: M
     integer, intent(in) :: n
     integer, intent(in) :: max_iter
     real(kind=rp), optional, intent(inout) :: rel_tol
     real(kind=rp), optional, intent(inout) :: abs_tol
+    logical, optional :: monitor
```
Comment (timfelle) [inline src/krylov/bcknd/sx/cg_sx.f90:70]:
```suggestion
    logical, optional, intent(in) :: monitor
```

**Review by MartinKarp — APPROVED** [review]
Good option I think! Can be merged as is I think. As for the output, are we considering to add the option to maybe only outputting for the first few time steps or to write out the residuals to a csv?

File: src/krylov/bcknd/sx/gmres_sx.f90 [hunk 10]
```diff
@@ -70,14 +70,16 @@ module gmres_sx
...
+  subroutine sx_gmres_init(this, n, max_iter, M, lgmres, &
+                           rel_tol, abs_tol, monitor)
     class(sx_gmres_t), intent(inout) :: this
     integer, intent(in) :: n
     integer, intent(in) :: max_iter
     class(pc_t), optional, intent(inout), target :: M
     integer, optional, intent(inout) :: lgmres
     real(kind=rp), optional, intent(inout) :: rel_tol
     real(kind=rp), optional, intent(inout) :: abs_tol
+    logical, optional :: monitor
```
Comment (timfelle) [inline src/krylov/bcknd/sx/gmres_sx.f90:82]:
```suggestion
    logical, optional, intent(in) :: monitor
```

File: src/krylov/bcknd/sx/pipecg_sx.f90 [hunk 11]
```diff
@@ -66,13 +66,14 @@ module pipecg_sx
...
   !> Initialise a pipelined PCG solver
-  subroutine sx_pipecg_init(this, n, max_iter, M, rel_tol, abs_tol)
+  subroutine sx_pipecg_init(this, n, max_iter, M, rel_tol, abs_tol, monitor)
     class(sx_pipecg_t), intent(inout) :: this
     class(pc_t), optional, intent(inout), target :: M
     integer, intent(in) :: n
     integer, intent(in) :: max_iter
     real(kind=rp), optional, intent(inout) :: rel_tol
     real(kind=rp), optional, intent(inout) :: abs_tol
+    logical, optional :: monitor
```
Comment (timfelle) [inline src/krylov/bcknd/sx/pipecg_sx.f90:76]:
```suggestion
    logical, optional, intent(in) :: monitor
```

**Review by timfelle — APPROVED** [review]

File: src/krylov/krylov.f90 (hunk 6 above)
Comment (njansson) [reply src/krylov/krylov.f90:318]: Sounds good , we can add it in another or

**Pull request merged**